### ISSUE TITLE
lv2lint errors

### ISSUE THREAD

**Issue by dvzrv** [surge-synthesizer/surge#2392]
**Describe the bug**
Hi! When building 1.7.0 as a package for Arch Linux I ran lv2lint against the resulting lv2 plugin.
Unfortunately lv2lint still fails on the plugin but with different issues than last time (#1698).

**Please let us know your surge version**

* Surge Version: 1.7.0
* Plugin type: lv2
* Bits: 64

**To Reproduce**

```
  export LDFLAGS="${LDFLAGS},-z,noexecstack"
  # build surge-headless
  cmake -DCMAKE_INSTALL_PREFIX='/usr' \
        -DCMAKE_BUILD_TYPE='None' \
        -W no-dev \
        -B build \
        -S .
  make VERBOSE=1 -C build

  # set datapath to local dir for testing
  sed -e 's|/usr/share/Surge|resources/data|' -i src/common/SurgeStorage.cpp
  # build surge-headless (test-suite
  cmake -DCMAKE_INSTALL_PREFIX='/usr' \
        -DCMAKE_BUILD_TYPE='None' \
        -W no-dev \
        -B build-test \
        -S .
  make VERBOSE=1 -C build-test
  lv2lint -Mpack -I "build-test/surge_products/${_name}.lv2" "https://surge-synthesizer.github.io/lv2/surge"
```

```
lv2lint 0.8.0
Copyright (c) 2016-2020 Hanspeter Portner (dev@open-music-kontrollers.ch)
Released under Artistic License 2.0 by Open Music Kontrollers
error: /build/surge/src/surge-1.7.0/build-test/surge_products/Surge.lv2/Surge_ui.ttl:23:0: unexpected end of file
lilv_world_load_file(): error: Error loading file `file:///build/surge/src/surge-1.7.0/build-test/surge_products/Surge.lv2/Surge_ui.ttl'
<https://surge-synthesizer.github.io/lv2/surge>
    [FAIL]  Plugin Symbols
              binary exports superfluous globally visible symbols:
                * configurationXmlStart
                * configurationXmlNullTerminator
                * lv2_generate_ttl
                * memorySVGListStart
              seeAlso: <http://lv2plug.in/ns/lv2core#binary>
    [WARN]  Plugin Author Email
              foaf:mbox not found
              seeAlso: <http://lv2plug.in/ns/lv2core#project>
  <https://surge-synthesizer.github.io/lv2/surge#UI>
    [FAIL]  UI Symbols
              binary exports superfluous globally visible symbols:
                * configurationXmlStart
                * configurationXmlNullTerminator
                * lv2_generate_ttl
                * memorySVGListStart
              seeAlso: <http://lv2plug.in/ns/lv2core#binary>
    [WARN]  UI Instance Access
              usage of instance-access is highly discouraged
              seeAlso: <http://lv2plug.in/ns/ext/instance-access>
    [WARN]  UI Mixed DSP/UI
              mixing DSP and UI code in same binary is discouraged
              seeAlso: <http://lv2plug.in/ns/extensions/ui#>
    [FAIL]  UI Idle Interface
              lv2:extensionData ui:idleInterface missing
              seeAlso: <http://lv2plug.in/ns/extensions/ui#idleInterface>
```

**Expected behavior**
lv2lint passes

**Screenshots**
n/a

**Desktop (please complete the following information):**
 - OS: Arch Linux
 - Host: n/a
 - Version: n/a

* gcc: 10.1.0
* cmake: 3.18.0
* lv2lint: 0.8.0

**Additional context**
lv2lint fails on different issues this time and lilv reports an `unexpected end of file`.

**Comment by baconpaul** [comment]
Definitely related to #2391 also

**Comment by baconpaul** [comment]
So those exported symbols

1: the configurationBlah ones are symbols we eject from assembly. Do you know how to make them non-exported in the resulting object? Basically we have a binary inclusion strategy where we generate

```
memorySVGListStart:
    .globl memorySVGListStart

    .incbin "/home/paul/dev/music/surge/assets/original-vector/SVG/exported/bmp00102.svg"
    .incbin "/home/paul/dev/music/surge/assets/original-vector/SVG/exported/bmp00105.svg"
    .incbin "/home/paul/dev/music/surge/assets/original-vector/SVG/exported/bmp00108.svg"
```

then refer to it in C++.

2: The lv2_generate_ttl is a symbol the LV2 exports so our python can run the C++ synth to generate the TLL file. I don't see a way to have it not exported. 

Thanks

**Comment by dvzrv** [comment]
Maybe @falkTX or @ventosus might know!

**Comment by baconpaul** [comment]
Yeah! Changing .globl to .local gave me a link error unsurprisingly.

**Comment by baconpaul** [comment]
So I just pushed #2393 which will solve all the problems except the exports.
I don't quite know how to deal with the exports. My gcc linker knowledge isn't quite up to snuff any more.

**Comment by ventosus** [comment]
@baconpaul, for lv2_generate_ttl, yes seem to need that to be globally exported, but you can whitelist it specifically in lv2lint:

    lv2lint [...] -s 'lv2_generate_ttl' [...]

**Comment by dvzrv** [comment]
With 1.8.1 the following is still the case:

```
lv2lint 0.10.0
Copyright (c) 2016-2021 Hanspeter Portner (dev@open-music-kontrollers.ch)
Released under Artistic License 2.0 by Open Music Kontrollers
<https://surge-synthesizer.github.io/lv2/surge>
    [FAIL]  Plugin Symbols
              binary exports superfluous globally visible symbols:
                * configurationXmlNullTerminator
                * configurationXmlStart
                * memorySVGListStart
                * lv2_generate_ttl
              seeAlso: <http://lv2plug.in/ns/lv2core#binary>
    [WARN]  Plugin Author Email
              foaf:mbox not found
              seeAlso: <http://lv2plug.in/ns/lv2core#project>
  <https://surge-synthesizer.github.io/lv2/surge#UI>
    [FAIL]  UI Symbols
              binary exports superfluous globally visible symbols:
                * configurationXmlNullTerminator
                * configurationXmlStart
                * memorySVGListStart
                * lv2_generate_ttl
              seeAlso: <http://lv2plug.in/ns/lv2core#binary>
    [WARN]  UI Instance Access
              usage of instance-access is highly discouraged
              seeAlso: <http://lv2plug.in/ns/ext/instance-access>
    [WARN]  UI Mixed DSP/UI
              mixing DSP and UI code in same binary is discouraged
              seeAlso: <http://lv2plug.in/ns/extensions/ui#>
```

I suspect, that I will have to whitelist `lv2_generate_ttl`, which is fine, but I believe it should be possible to fix the other ones.

**Comment by baconpaul** [comment]
We are defacto not supporting the lv2 at current and will retire it after the 19 release when we move to juce.

**Comment by dvzrv** [comment]
Oh. I was not aware. That's sad to hear! :-/

**Comment by baconpaul** [comment]
VST3 works great on linux. 
also falk said once we have juce he would try a build with his fork
But the lv2 in the code base doesn’t work correctly and hasn’t been fixed by anyone, and basically juce is a way better way to write a plugin

**Comment by diovudau** [comment]
No LV2 is sad to hear.
It is the backbone for Linux Audio plugins.

What are the technical difficulties? Now that surge is open source the solution is not dependent on a few people alone anymore.

**Comment by baconpaul** [comment]
All the issues are here in the lv2 target.

But basically surge dsp engine can change params on automation. That’s never worked reliably in the lv2 spec

But I’m not sure what’s wrong with VST3. It works in every linux environment at this point and is GPL3.

the real thing we want is to remove dependency on any hand written wrapper and let the juce layer do It all though. That’s the change coming. And then juce supporting lv2 becomes the work.

**Comment by dvzrv** [comment]
> But I’m not sure what’s wrong with VST3. It works in every linux environment at this point and is GPL3.

Well, I guess some people fear, that due to the dual-license and Steinberg [effectively not letting outside collaborators merge in changes](https://github.com/steinbergmedia/vst3sdk/graphs/contributors) means one thing: The license will be reverted once it has ensured that vst3 has reached a large enough adoption.

>  And then juce supporting lv2 becomes the work.

This will very likely only ever happen in an outside-of-juce effort though, judging from their standpoint: https://github.com/juce-framework/JUCE/issues/123
With JUCE (and whichever the current owner is), the concerns are probably similar to those towards Steinberg in regards to code licensing.

Therefore I am glad, that a fork with lv2 support exists! :)
If this is the easier solution for surge, then that's  probably a good choice for maintainability!

**Comment by unfa** [comment]
Hmm, so after switch to JUCE we may get LV2 back? AFAIK @falkTX has his own JUCE fork that supports LV2.

I am not super happy about VST3, mostly due to it coming from Steinberg, who have forced everyone to drop VST2. It's great to see it being GPL3 licensed, though not all FOSS audio software supports it fully yet, and I think FOSS audio developers usually regard LV2 as technically superior. I'm not a programmer so I can't say why exactly. I know I prefer LV2 whenever I can use it, though sadly Surge so far was pretty unusable in LV2, and I've been using the VST3 version by necessity.
I'd love to see LV2 go back, though maybe it's just NIH syndrome, I'm biased.

**Comment by mxmilkiib** [comment]
Indeed, Surge could consider https://github.com/DISTRHO/JUCE after the heavy lifting to JUCE itself is done, maybe?

**Comment by baconpaul** [comment]
Yeah falk said he would try it. The mainline code now has a surge which is starting to work in juce too.

I understand not loving the VST3 api. I want to dump coding directly to it. It has been hard. But lv2 has been as hard if not harder so far. This is why juce is appealing

of course if you disagree I will tag the release point where we drop hand rolled plugins so you could always fork from there as a starting point if juce lv2 doesn’t work for you!

**Comment by falkTX** [comment]
the lv2 version should not really be a problem.
lets wait until there is an official surge release with juce as its api (even if it is a beta), and then go from there trying out the lv2 version.

Worth noting is that the DISTRHO JUCE fork is a drop-in replacement.

@baconpaul surge will use juce cmake stuff for building, correct? I can focus my attention there, so it will be easier to handle later on. There is already a PR for adding lv2 support for juce cmake stuff (in the DISTRHO fork, of course)

**Comment by baconpaul** [comment]
@falkTX there’s an alpha release right now in the codebase and even though it has quite a few things which don’t work or will change, it is buildable. We actually distribute the binary as a separate zip in the nightlies.

To build it, checkout and stage submodules then

```
cmake -Bbuildjuce -DBUILD_SURGE_JUCE_PLUGINS=TRUE -DCMAKE_BUILD_TYPE=Debug
cmake —build build —config Debug —target surge-xt_All
```

You will see that right now I haven’t staged juce as a submodule so the above does a checkout of juce explicitly with git. (I will change this after then 19 release, when I add juce as a submodule and lose vstgui and vst3sdk) so you gotta do at least stage one on an internet connected box with git. 

The cmake rules are really pretty easy. Just look for the surge-xt stuff

Happy to add or merge a SURGE_WITH_DISTRO_JUCE=(whatever) set of cmake changes right now. Our strategy is to basically run a juce alpha side by side with the 19 development then once 19 is done retire the non-juce plugins and push to the surge juce. This is all summarized in one of the pinned issues here.

Oh also, we are going to change the plugin name and even though we will maintain patch compatibility will not retain daw automation compatibility. So surge 1.9 and surge-xt 1.0 will be able to run side by side as separate plugs when we are done.

Finally right now surge-xt requires the same staged assets as surge in /usr/share/surge etc.

**Comment by falkTX** [comment]
Nice. I do not have time to check on this right now, have other priorities at the moment, but I want to check on this once it reaches beta state.
From what you write, it seems like to me that adding an lv2 version will be pretty straight-forward. I just need to ensure the juce lv2 wrapper works correctly once we reach that point.

**Comment by mkruselj** [comment]
> Well, I guess some people fear, that due to the dual-license and Steinberg [effectively not letting outside collaborators merge in changes](https://github.com/steinbergmedia/vst3sdk/graphs/contributors) means one thing: The license will be reverted once it has ensured that vst3 has reached a large enough adoption.

I kinda don't think they'd do that.

**Comment by baconpaul** [comment]
> > Well, I guess some people fear, that due to the dual-license and Steinberg [effectively not letting outside collaborators merge in changes](https://github.com/steinbergmedia/vst3sdk/graphs/contributors) means one thing: The license will be reverted once it has ensured that vst3 has reached a large enough adoption.
> 
> I kinda don't think they'd do that.

while they could definitely write software that isn't GPL3, there's no way for their current GPL3 software to remove that license. So they could best say "features going forward are not GPL3".

**Comment by baconpaul** [comment]
Oh and to the two folks who thumbed down my comment that “we are defacto not supporting the LV2” - there is a way for you to fix that! Dive in and support it. Close the issues here. Subscribe to the github and answer support requests. And so on. We welcome anyone who wants to contribute. 

But no one active on the surge project right now is working on, uses, regularly writes music using daws that run, or has any experience with LV2. Our plan is to hope that @falkTX ’s future juce fork will allow it to work better. 

But rather than thumbs downing comments on github, fire up an IDE and fix the things you want! That’s how every other feature in surge got here (and it’s how the LV2 got here too!)  Happy to merge changes.

**Comment by baconpaul** [comment]
anyway @falkTX over in #4004 I added a cmake-time option to use DISTRHO-JUCE in the SurgeEffectsBank production build and Surge-XT experimental build. If you do

```
cmake -Bwhatever -DBUILD_SURGE_JUCE_PLUGINS=True -DBUILD_SURGE_WITH_DISTRHO_JUCE=True
cmake --build whatever --config Release --target surge-xt_Standalone
```

it will build XT. The production version of the FX bank builds using this mechanism now (target `surge-fx_VST3` for instance)

I tried to activate the LV2 by adding a "FORMATS LV2" to the master CMakeLists.txt but that failed to generate a `surge-fx_LV2` target when CMake was re-run so I think you may need to tweak your branch to do the right thing in that build. Looks like you have some time in `extras/Build/CMake/JUCEUtils.cmake` in your future, copying the VST3 code path there and adding an LV2 equivalent. But from looking at your code, looks tractable. And all the standalone built with your branch seem to work just fine.

Also you don't seem to have the same tag/branch strategy as juce. I couldn't find a 6.0.7 tag in your repo so I just picked main.

But if you can make distrho build the fx bank, I can include a production LV2 of that with the 1.9 release in april since it already is a prod surge juce plugin. If you want to do that though we should probably open a separate issue over here.

**Comment by baconpaul** [comment]
Finally over in https://github.com/DISTRHO/JUCE/issues/14 I shared some cmake diffs I used to make at least the .so of the LV2 link for surge-xt with cmake. No ttl or anything else though. Still work to do. but I bet you can make the FXbank LV2 work pretty quickly from the surge change here and that diff pls a bit of tweaking.

**Comment by baconpaul** [comment]
Closing this issue, since #4030 is the place we are investing in LV2 going forward. If the JUCE/LV2 port works, it works, and the problems in our hand-port will be gone (or replaced with problems in the JUCE/LV2 port).
